src/stest/src/file.rs

```diff
@@ -175,6 +173,13 @@ impl File {
         self.mode().map(is_executable)
     }
 
+    pub fn file_name(&self) -> String {
+        self.path_buf
+            .file_name()
+            .map(|f| f.to_string_lossy().to_string())
+            .unwrap_or_else(|| "".to_string())
+    }
```

Comment:
I like how the code in this PR, in contrast to https://github.com/Shizcow/dmenu-rs/pull/54, leaves the decision of how `stest` should format its output (whether `stest` should output either a file name or the full path of a file) to the [conditional][1] in the `write` function. The `write` function feels like the right place to make decisions about how to format the output of `stest`. It's more toward the "outer shell" of the application, rather than the "inner core".

In a similar vein, I think we should go further and move more decisions out of `File`'s interface and into the `write` function. For example, how to convert a `File -> String` is probably best left to the `write` function. 

We can implement this by having `file_name()` return a `File` instead of a `String`.

```suggestion
    pub fn file_name(&self) -> Option<File> {
        self.path_buf
            .file_name()
            .map(|os_str| PathBuf::from(os_str))
            .map(|path_buf| File::new(path_buf))
    }
```

This also leaves the decision of how to handle `None` cases ([where a file name is requested for a file path that terminates in `..`][2]) to the caller.

I think this gives `File` a cleaner interface. It also makes it easier to fix the broken test by adding code like

```rust
                .map(|file| file.file_name().unwrap())
```

between [these two lines][3].

[1]: https://github.com/Shizcow/dmenu-rs/pull/55/files#diff-189e6556964731ef3ba769944566a34e8d1af475f80a56dcc611bf19bc9c5addR159
[2]: https://doc.rust-lang.org/std/path/struct.PathBuf.html#method.file_name
[3]: https://github.com/Shizcow/dmenu-rs/blob/b2a61fbdee7cb7c1a8eec9f16edabd43f00d58e8/src/stest/tests/integration_tests.rs#L116-L117

Reply:
i see. that makes sense. thanks for taking the time for a detailed explanation.
i'll check the test and update.

Reply:
updated with your suggestions i've fixed the failing test, thanks for pointing it out.
i've also updated the name of this method to `clone_with_path_as_file_name` , let me know if you have a better suggestion.

Reply:
I think `file_name` is a good name for the method.

Unless there is some important changes in ownership that occur in the function that `clone_with` better describes? Either are fine.

Reply:
i guess `file.file_name()` kinda implies to me that it's returning an attribute of the file (i.e. its name as a string), hence went with the more verbose `clone_with_path_as_file_name` which expresses that the return type is going to be a `File`.

Reply:
Ah, I can see that.

I personally don't find `file_name` a confusing name for this method, because I understand `File` to mean unix path in this context, and the file name component of a unix path is always itself a valid unix path.

But I think both names are serviceable for this small utility method.